File: src/contracts/status.c
#include "usdk/status.h"

const char* usdk_status_string(usdk_status_t status) {
    switch (status) {
        case USDK_OK: return "ok";
        case USDK_ERR_INVALID_ARGUMENT: return "invalid argument";
        case USDK_ERR_ABI_VERSION_MISMATCH: return "ABI version mismatch";
        case USDK_ERR_STRUCT_SIZE_MISMATCH: return "struct size mismatch";
        case USDK_ERR_CAPABILITY_NOT_FOUND: return "capability not found";
        case USDK_ERR_MODULE_LOAD_FAILED: return "module load failed";
        case USDK_ERR_MODULE_ENTRY_NOT_FOUND: return "module entry point not found";
        case USDK_ERR_DEPENDENCY_CYCLE: return "dependency cycle detected";
        case USDK_ERR_DEPENDENCY_UNRESOLVED: return "dependency unresolved";
        case USDK_ERR_DEPENDENCY_VERSION_INCOMPATIBLE: return "dependency version incompatible";
        case USDK_ERR_STALE_ROUND: return "stale or replayed round";
        case USDK_ERR_DUPLICATE_VOTE: return "duplicate vote";
        case USDK_ERR_PARTY_CONFLICT: return "party identity conflict";
        case USDK_ERR_CANDIDATE_MUTATED: return "candidate mutated after vote";
        case USDK_ERR_DEADLINE_EXCEEDED: return "deadline exceeded";
        case USDK_ERR_ROUND_NOT_OPEN: return "round not open";
        case USDK_ERR_OUT_OF_MEMORY: return "out of memory";
        case USDK_ERR_IO: return "I/O error";
        case USDK_ERR_NOT_IMPLEMENTED: return "not implemented";
        default: return "unknown status";
    }
}

File: packages/contracts/src/status.mjs
/**
 * Status codes, mirroring usdk_status_t (include/usdk/status.h) in the
 * native C SDK - not a strict 1:1 copy (JS has fewer categories worth
 * distinguishing, e.g. no separate struct-size-mismatch concept), but
 * intentionally overlapping names so a developer moving between the C
 * ABI and this JS layer recognizes the vocabulary.
 */
export const Status = Object.freeze({
  OK: 'ok',
  INVALID_ARGUMENT: 'invalid-argument',
  ABI_VERSION_MISMATCH: 'abi-version-mismatch',
  CAPABILITY_NOT_FOUND: 'capability-not-found',
  MODULE_LOAD_FAILED: 'module-load-failed',
  DEPENDENCY_CYCLE: 'dependency-cycle',
  DEPENDENCY_UNRESOLVED: 'dependency-unresolved',
  DEPENDENCY_VERSION_INCOMPATIBLE: 'dependency-version-incompatible',
  STALE_ROUND: 'stale-round',
  DUPLICATE_VOTE: 'duplicate-vote',
  PARTY_CONFLICT: 'party-conflict',
  CANDIDATE_MUTATED: 'candidate-mutated',
  DEADLINE_EXCEEDED: 'deadline-exceeded',
  ROUND_NOT_OPEN: 'round-not-open',
  NOT_IMPLEMENTED: 'not-implemented',
});

/** Thrown by @usdk/* functions instead of returning a bare error string,
 * so callers can `catch (e) { if (e instanceof UsdkError) ... }` and
 * still read a stable `e.status` code from the Status enum above. */
export class UsdkError extends Error {
  /** @param {string} status one of the Status values @param {string} message */
  constructor(status, message) {
    super(message);
    this.name = 'UsdkError';
    this.status = status;
  }
}
